File: src/io_stream/io_stream_ffi.mjs
import {
    Result$Ok,
    Result$Ok$0,
    Result$Error,
    Result$Error$0,
    Result$isError,
    BitArray$BitArray,
    BitArray$BitArray$data,
} from "../gleam.mjs";
import fs from "fs";

const utf8Decoder = new TextDecoder("utf-8", { fatal: true });
const utf8Encoder = new TextEncoder();

class Handle {
    #fd;
    #position;

    constructor(fd) {
        this.#fd = fd;
        this.#position = 0;
    }

    read(count) {
        const buffer = Buffer.allocUnsafe(count);
        const bytesRead = fs.readSync(
            this.#fd,
            buffer,
            0,
            count,
            this.#position,
        );

        this.#position += bytesRead;

        return buffer.subarray(0, bytesRead);
    }

    write(buffer) {
        const bytesWritten = fs.writeSync(
            this.#fd,
            buffer,
            0,
            buffer.byteLength,
            this.#position,
        );

        this.#position += bytesWritten;
    }

    seek(position) {
        this.#position = position;
    }

    sync() {
        fs.fsyncSync(this.#fd);
    }

    close() {
        fs.closeSync(this.#fd);
    }
}

// Open operations

export function open(path, modes) {
    const flags = translateModes(modes);

    try {
        const fd = fs.openSync(path, flags);
        return Result$Ok(new Handle(fd));
    } catch (error) {
        return errorWithDesc(error);
    }
}

function translateModes(modes) {
    const has = (mode) => {
        let list = modes;
        while (list && list.head) {
            let curr = list.head.constructor.name;
            if (curr == mode) return true;
            list = list.tail;
        }
        return false;
    };

    const read = has("Read");
    const write = has("Write");
    const append = has("Append");
    const exclusive = has("Exclusive");
    const truncate = has("Truncate");

    let flag;

    if (read && write) {
        if (truncate) {
            // w+ : read/write, truncate, create
            flag = "w+";
        } else {
            // r+ : read/write, preserve
            flag = "r+";
        }
    } else if (append) {
        // a = append + create
        flag = "a";

        if (exclusive) {
            flag = "ax";
        }
    } else if (write) {
        // w = write + create + truncate
        flag = "w";

        if (exclusive) {
            flag = "wx";
        }
    } else if (read) {
        // r = read existing
        flag = "r";
    } else {
        throw new Error("Invalid open mode");
    }

    if (has("Binary")) {
        // Node has no binary flag.
        // Buffers are returned by default when encoding is not specified.
    }

    return flag;
}

// Read operations

export function next_byte(handle) {
    try {
        const buffer = handle.read(1);

        if (buffer.byteLength === 0) {
            return endOfFileError();
        }

        return Result$Ok(buffer[0]);
    } catch (error) {
        return errorWithDesc(error);
    }
}


export function read_bytes(handle, count) {
    try {
        const buffer = handle.read(count);

        if (buffer.byteLength === 0) {
            return endOfFileError();
        }

        return Result$Ok(BitArray$BitArray(buffer));
    } catch (error) {
        return errorWithDesc(error);
    }
}

export function next_char(handle) {
    const buffer = handle.read(1);

    if (buffer.byteLength == 0) {
        return endOfFileError();
    }

    const b0 = buffer[0];

    let len;

    if ((b0 & 0x80) === 0x00) {
        len = 1;
    } else if ((b0 & 0xe0) === 0xc0) {
        len = 2;
    } else if ((b0 & 0xf0) === 0xe0) {
        len = 3;
    } else if ((b0 & 0xf8) === 0xf0) {
        len = 4;
    } else {
        return errorWithDesc({
            code: "INVALID_UTF8",
            message: "Invalid UTF-8 leading byte"
        })
    }

    // Read remaining bytes
    if (len > 1) {
        const rest = handle.read(len - 1)
        if (rest.byteLength !== len - 1) {
            return errorWithDesc({
                code: "INVALID_UTF8",
                message: "Unexpected EOF while reading UTF-8 character"
            })
        }
        buffer = Buffer.concat([buffer, rest])
    }

    try {
        let char = utf8Decoder.decode(buffer);
        return Result$Ok(char);
    }
    catch (err) {
        return errorWithDesc({
            code: "INVALID_UTF8",
            message: err.message
        })
    }
}


export function read_line(handle) {
    try {
        let output = "";

        while (true) {
            const char = next_char(handle);

            if (Result$isError(char)) {
                if (Result$Error$0(char)[0] === "EOF") {
                    return Result$Ok(output);
                }
                return char;
            }

            output += Result$Ok$0(char);

            if (output.charAt(output.length - 1) == "\n") {
                return Result$Ok(output);
            }
        }
    } catch (error) {
        return errorWithDesc(error);
    }
}


// Write operations

export function write_bytes(handle, bytes) {
    let data = BitArray$BitArray$data(bytes);

    try {
        handle.write(data);
        return Result$Ok(undefined);
    } catch (error) {
        return errorWithDesc(error);
    }
}


export function write_string(handle, string) {
    let data = utf8Encoder.encode(string)
    try {
        handle.write(data);
        return Result$Ok(undefined);
    } catch (error) {
        return errorWithDesc(error);
    }
}


export function write_line(handle, string) {
    let data = utf8Encoder.encode(string + "\n")
    try {
        handle.write(data);
        return Result$Ok(undefined);
    } catch (error) {
        return errorWithDesc(error);
    }
}


// Generic operations

export function close(handle) {
    try {
        handle.close();

        return Result$Ok(undefined);
    } catch (error) {
        return errorWithDesc(error);
    }
}


export function seek(handle, position) {
    try {
        handle.seek(position, 0);

        return Result$Ok(undefined);
    } catch (error) {
        return errorWithDesc(error);
    }
}

export function sync(handle) {
    try {
        if (handle.sync) {
            handle.sync();
        }

        return Result$Ok(undefined);
    } catch (error) {
        return errorWithDesc(error);
    }
}

// Standard streams

export function stdin() {
    return new Handle(process.stdin.fd);
}

export function stdin_bin() {
    return new Handle(process.stdin.fd);
}


export function stdout() {
    return new Handle(process.stdout.fd);
}


export function stdout_bin() {
    return new Handle(process.stdout.fd);
}


export function stderr() {
    return new Handle(process.stderr.fd);
}


export function stderr_bin() {
    return new Handle(process.stderr.fd);
}

// Error helper functions

function errorWithDesc(e) {
    return Result$Error([e.code ?? "UNKNOWN", e.message ?? "Unknown error"]);
}

function endOfFileError() {
    return errorWithDesc({ code: "EOF", message: "End of file reached" });
}

File: src/io_stream/io_stream_ffi.erl
-module(io_stream_ffi).

-export([
    open/2,

    next_byte/1,
    read_bytes/2,
    next_char/1,
    read_line/1,

    write_bytes/2,
    write_string/2,
    write_line/2,

    close/1,
    seek/2,
    sync/1,

    stdin/0,
    stdin_bin/0,
    stdout/0,
    stdout_bin/0,
    stderr/0,
    stderr_bin/0
]).

open(Path, Modes) ->
    ErlangModes = translate_modes(Modes),
    case file:open(Path, ErlangModes) of
        {ok, IoDevice} ->
            {ok, IoDevice};

        {error, Reason} ->
            {error, error_with_desc(Reason)}
    end.

%% Mode translation

translate_modes(Modes) ->
    lists:reverse(
        lists:foldl(fun translate_mode/2, [], Modes)
    ).

translate_mode('read', Acc) ->
    [read | Acc];

translate_mode('write', Acc) ->
    [write | Acc];

translate_mode('append', Acc) ->
    [append | Acc];

translate_mode('binary', Acc) ->
    [binary | Acc];

translate_mode('truncate', Acc) ->
    [truncate | Acc];

translate_mode('exclusive', Acc) ->
    [exclusive | Acc].

%% Read operations

next_byte(IoDevice) ->
    case file:read(IoDevice, 1) of
        {ok, <<Byte>>} ->
            {ok, Byte};

        eof ->
            {error, {<<"EOF">>, "Reached End of file"}};

        {error, Reason} ->
            {error, error_with_desc(Reason)}
    end.

read_bytes(IoDevice, Count) ->
    case file:read(IoDevice, Count) of
        {ok, Bytes} ->
            {ok, Bytes};

        eof ->
            {error, {<<"EOF">>, "Reached End of file"}};

        {error, Reason} ->
            {error, error_with_desc(Reason)}
    end.

next_char(IoDevice) ->
    case io:get_chars(IoDevice, "", 1) of
        eof ->
            {error, {<<"EOF">>, "Reached End of file"}};

        {error, Reason} ->
            {error, error_with_desc(Reason)};

        Chars when is_list(Chars) ->
            {ok, unicode:characters_to_binary(Chars)}
    end.

read_line(IoDevice) ->
    case io:get_line(IoDevice, "") of
        eof ->
            {error, {<<"EOF">>, "Reached End of file"}};

        {error, Reason} ->
            {error, error_with_desc(Reason)};

        Line when is_list(Line) ->
            {ok, unicode:characters_to_binary(Line)};

        Line when is_binary(Line) ->
            {ok, Line}
    end.

%% Write operations

write_bytes(IoDevice, Bytes) ->
    case file:write(IoDevice, Bytes) of
        ok ->
            {ok, nil};

        {error, Reason} ->
            {error, error_with_desc(Reason)}
    end.

write_string(IoDevice, String) ->
    case file:write(IoDevice, unicode:characters_to_binary(String)) of
        ok ->
            {ok, nil};

        {error, Reason} ->
            {error, error_with_desc(Reason)}
    end.

write_line(IoDevice, String) ->
    case file:write(
        IoDevice,
        [unicode:characters_to_binary(String), <<"\n">>]
    ) of
        ok ->
            {ok, nil};

        {error, Reason} ->
            {error, error_with_desc(Reason)}
    end.

%% Generic operations

close(IoDevice) ->
    case file:close(IoDevice) of
        ok ->
            {ok, nil};

        {error, Reason} ->
            {error, error_with_desc(Reason)}
    end.

seek(IoDevice, Position) ->
    case file:position(IoDevice, Position) of
        {ok, _} ->
            {ok, nil};

        {error, Reason} ->
            {error, error_with_desc(Reason)}
    end.


sync(IoDevice) ->
    case file:sync(IoDevice) of
        ok ->
            {ok, nil};

        {error, Reason} ->
            {error, error_with_desc(Reason)}
    end.

%% Standard streams

stdin() ->
    standard_io.

stdin_bin() ->
    standard_io.

stdout() ->
    standard_io.

stdout_bin() ->
    standard_io.

stderr() ->
    standard_error.

stderr_bin() ->
    standard_error.

%% Helper function for errors
error_with_desc(Reason) when is_atom(Reason) ->
    {
        unicode:characters_to_binary(atom_to_list(Reason)),
        unicode:characters_to_binary(file:format_error(Reason))
    }.